File: Complaint Redressal System/source code/frontend/src/app/services/complaint.service.ts
import { Injectable } from '@angular/core';
import { Complaint } from '../classes/complaint';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from '../classes/user';

@Injectable({
  providedIn: 'root'
})
export class ComplaintService {
  private apiUrl ="http://localhost:8080/complaints";

  constructor(private http:HttpClient) { }

  saveComplaint(complaint:Complaint) 
  {
    return this.http.post(`${this.apiUrl}`,complaint);
  }
  getAllComplaints() : Observable<Complaint[]>
  {
    return this.http.get<Complaint[]>(`${this.apiUrl}`);
  }
  
  getComplaintsByUser(user:string): Observable<Complaint[]>
  {
    return this.http.get<Complaint[]>(`${this.apiUrl+"/user"}/${user}`);
  }
  updateComplaint(id: number, value: String)
  {
    return this.http.put(`${this.apiUrl}/${id}`,value);
  }
  getComplaintsByEngineerAssigned(engineer:string): Observable<Complaint[]>
  {
    return this.http.get<Complaint[]>(`${this.apiUrl+"/engineer"}/${engineer}`);
  }
}

File: Complaint Redressal System/source code/backend/src/main/java/com/example/demo/service/ComplaintService.java
package com.example.demo.service;

import java.util.List;

import org.springframework.http.ResponseEntity;

import com.example.demo.Model.Complaint;

public interface ComplaintService {

	public ResponseEntity<String> addComplaint(Complaint complaint);
	
	public List<Complaint> geAllComplaints();
	
	public List<Complaint> getComplaintsByUser(String user);
	
	public ResponseEntity<String> updateComplaint(int id, String complaint);
	
	public List<Complaint> getComplaintsByEngineerAssigned(String engineer);
}
